### ISSUE TITLE
how to release memory after playing video

### ISSUE THREAD

**Issue by rikensonaniNB** [RenderHeads/UnityPlugin-AVProVideo#1315]
**Describe the issue**
how to release memory after playing video. i am try with CloseVideo but memory not releasing. also i am pick video from storage and playing then repick another video and playing etc. every time increase memory not clearing if you have any solution then let me know.

also i am try to use mediaPlayer.ForceDispose(); but not releasing memory and getting null reference exception please check attachment screenshot.
also we can't use System.GC.Collect(); because it's very expensive.

**Your Setup (please complete the following information):**
 - Unity version: 2021.3.4f1
 - AVPro Video version (number and edition (trial/core/ultra/enterprise)): 2.4.1(plugin v2.4.1-ultra)
 - Operating system version: 
 - Device model: windows 10
 - Video specs (resolution, frame-rate, codec, file size): not specific

**Comment by Chris-RH** [comment]
1. How are you measuring the memory?
2. How much is the memory increasing each time you are loading a video?
3. Is it continuous or does it gradually increase?
4. Are you loading a new video in the same mediaplayer component or  is it loading in a second one while the first video is playing?
5. When you close the mediaplayer component, does the memory decrease at all? How long have you left it?

**Comment by rikensonaniNB** [comment]
1)checking in XCode memory profile 
2)not fixed but 5 to 15mb increase 
3)it's every time increase pause after play 
4)i have created lots of item like 10 video url then create 10 item in every item that own media player once item on screen then this item media player is playing other all is pause then after i swipe top to bottom then next item is on screen then this item media player play and last one is on pause 
5)not sure memory is decrease or not. approximate minor memory decrease

**Comment by Chris-RH** [comment]
Could you update to the latest version please and let us know if you are still having the same problem?
The memory won't decrease after pausing, only after disposing of the mediaplayer.

**Comment by rikensonaniNB** [comment]
ok let me try with latest version. if fetching same issue then let you know
thank you

**Comment by rikensonaniNB** [comment]
hello @Chris-RH i have update this package and i am trying to creating android build getting this error. so can i delete or ignore?

![image (1)](https://user-images.githubusercontent.com/87807293/194557918-e98ba976-efb9-4230-87e7-83bb94039eac.png)

**Comment by Ste-RH** [comment]
Click ‘delete’

**Comment by rikensonaniNB** [comment]
> also i have one Quotation if media player is playing and media player is pause both have any different for memory allocation and performance ?

could you give me replay as well

**Comment by myeg** [comment]
I also found that some textures remained in GPU after I called mediaPlayer.Stop() , mediaPlayer.CloseMedia() and mediaPlayer.ForceDispose();  It cannot be released. I captured one frame from Xcode, the result is shown in below picture
![image](https://user-images.githubusercontent.com/19905704/195247770-80f81283-61f4-4177-9686-516c8bda9ba2.png)

**Comment by Chris-RH** [comment]
@myeg,
Are you using the latest version, 2.6.4?
We recently fixed an issue on macOS/iOS/tvOS with textures persisting in the texture cache after all player instances have been released.

**Comment by rikensonaniNB** [comment]
> I also found that some textures remained in GPU after I called mediaPlayer.Stop() , mediaPlayer.CloseMedia() and mediaPlayer.ForceDispose(); It cannot be released. I captured one frame from Xcode, the result is shown in below picture

@Chris-RH could you give me replay please
i am using latest version

**Comment by rikensonaniNB** [comment]
> > also i have one Quotation if media player is playing and media player is pause both have any different for memory allocation and performance ?
> 
> could you give me replay as well

@Chris-RH ?

**Comment by rikensonaniNB** [comment]
some time same video play with url (open with Auto play false)
and some time same video don't play with url and returning this error (open with Auto play)

[AVProVideo] Error: Loading failed.  File not found, codec not supported, video resolution too high or insufficient system resources.

**Comment by rikensonaniNB** [comment]
also i load multiple video at same time and play one video app is crash on android. here is crash log please check if you have any solution then let me know.  

E/AndroidRuntime: FATAL EXCEPTION: ExoPlayer:Playback
    java.lang.OutOfMemoryError: Failed to allocate a 184 byte allocation with 0 free bytes and 0B until OOM, target footprint 536870912, growth limit 536870912
    
 here is some related discussion    
![image](https://user-images.githubusercontent.com/87807293/195585674-eec0b99d-b392-482a-8492-220a51cdffe7.png)

**Comment by Ste-RH** [comment]
Please report different issues as their own issue.  We cannot keep track of multiple issues under a single issue.

**Comment by Ste-RH** [comment]
> also i load multiple video at same time and play one video app is crash on android. here is crash log please check if you have any solution then let me know.
> 
> E/AndroidRuntime: FATAL EXCEPTION: ExoPlayer:Playback java.lang.OutOfMemoryError: Failed to allocate a 184 byte allocation with 0 free bytes and 0B until OOM, target footprint 536870912, growth limit 536870912
> 
> here is some related discussion ![image](https://user-images.githubusercontent.com/87807293/195585674-eec0b99d-b392-482a-8492-220a51cdffe7.png)

Your Device is out of memory.  It is highly unlikely this is due to AVPro Video...unless there is a specific memory leak you can point at on Android?

We run stress tests where we open/close videos some 50,000 times to check for leaks.  I ran this test just yesterday on the last release (2.6.4) and it passed without issue.

**Comment by rikensonaniNB** [comment]
> > also i load multiple video at same time and play one video app is crash on android. here is crash log please check if you have any solution then let me know.
> > E/AndroidRuntime: FATAL EXCEPTION: ExoPlayer:Playback java.lang.OutOfMemoryError: Failed to allocate a 184 byte allocation with 0 free bytes and 0B until OOM, target footprint 536870912, growth limit 536870912
> > here is some related discussion ![image](https://user-images.githubusercontent.com/87807293/195585674-eec0b99d-b392-482a-8492-220a51cdffe7.png)
> 
> Your device is out of memory. It is highly unlikely this is due to AVPro Video...unless there is a specific memory leak you can point at on Android?
> 
> We run stress tests where we open/close videos some 50,000 times to check for leaks. I ran this test just yesterday on the last release (2.6.4) and it passed without issue.

1)this issue occurring only on android and i have setup screen with video same as Instagram reals screen. i think this issue occurring after update latest version.  

2)also this issue more times occurring after update to latest version.
some time same video play with url (open with Auto play false)
and some time same video don't play with url and returning this error (open with Auto play)

[AVProVideo] Error: Loading failed. File not found, codec not supported, video resolution too high or insufficient system resources.

sorry for more asking but if you found any issue then let me know it's very important to me.

**Comment by Ste-RH** [comment]
Android devices vary in how much memory they have with cheaper devices having a lot less RAM available to applications compared to, say, most iOS devices.

1) What is your device?
2) How many videos have you got concurrently open?
3) What version of AVPro Video are you running, and have you tested the same scenario on the same device using a previous version (that you say worked)?  What is the version is that previous version?
4) The video that sometimes fails to open.  Is this coming off a cloud server, or is it local?  Can you provide us with the video or the URL to it?

**Comment by rikensonaniNB** [comment]
1)Redmi note 8 pro (6GB Ram/ 64GB Storage) 10GB Storage free now
2)10 video load with auto play false. then open full screen view like realse screen in this screen we play one video is visible on screen other then all is pause and if current video swipe up and down more then 10 we close this video for memory releasing and is item back to less then 10 we load video with auto play false
3)Current is Version 2.6.4 (plugin v2.6.3f1-ultra) and previous version is 2.4.1(plugin v2.4.1-ultra)
4)video coming from aws server here is url
=>https://cdn.xana.net/apitestxana/Defaults/1661163273635-MediaFile.mp4
=>https://cdn.xana.net/apitestxana/Defaults/1662967639450-pickedMedia1.mp4
=>https://cdn.xana.net/apitestxana/Defaults/1652172793013_recording_2022_05_10_14_22_37_508.mp4

i am load with auto play false then same url load with thumbnail success but i am trying same video with auto play true then some time returning error above i share you. other then working fine

**Comment by Dejavu0709** [comment]
ios crash because of memory when multi(4) videos are playing

**Comment by myeg** [comment]
> @myeg, Are you using the latest version, 2.6.4? We recently fixed an issue on macOS/iOS/tvOS with textures persisting in the texture cache after all player instances have been released.

Yes, I have updated to the latest version 2.6.4, but the textures still can not be released after destroying all media players.

**Comment by rikensonaniNB** [comment]
> 1)Redmi note 8 pro (6GB Ram/ 64GB Storage) 10GB Storage free now 2)10 video load with auto play false. then open full screen view like realse screen in this screen we play one video is visible on screen other then all is pause and if current video swipe up and down more then 10 we close this video for memory releasing and is item back to less then 10 we load video with auto play false 3)Current is Version 2.6.4 (plugin v2.6.3f1-ultra) and previous version is 2.4.1(plugin v2.4.1-ultra) 4)video coming from aws server here is url =>https://cdn.xana.net/apitestxana/Defaults/1661163273635-MediaFile.mp4 =>https://cdn.xana.net/apitestxana/Defaults/1662967639450-pickedMedia1.mp4 =>https://cdn.xana.net/apitestxana/Defaults/1652172793013_recording_2022_05_10_14_22_37_508.mp4
> 
> i am load with auto play false then same url load with thumbnail success but i am trying same video with auto play true then some time returning error above i share you. other then working fine
> 
> 5)also i open video with auto play false or auto play true both are allocate same memory ?

@Ste-RH @Chris-RH please replay asap.
thank you

**Comment by Ste-RH** [comment]
@rikensonaniNB 10 videos open all at once is quite a lot to ask an Android device.  Opening with or without AutoPlay enabled makes no difference.  When you open a video the underlying system needs to allocate resources and decode the first frame(s) ready to render right away.

You could lower your video resolutions.  That might help.  The only other thing that might help is if you set some videos to be software decoded.  Even then, resources will need to be allocated, but this should spread over the GPU memory and ram a little better.

The bottom line is, you are opening too many videos concurrently.  Your app will for sure fail on lesser devices with lower hardware specifications.

**Comment by rikensonaniNB** [comment]
> 2)also this issue more times occurring after update to latest version. some time same video play with url (open with Auto play false) and some time same video don't play with url and returning this error (open with Auto play)
> 
> [AVProVideo] Error: Loading failed. File not found, codec not supported, video resolution too high or insufficient system resources.

thank you for support 
what about this error ?
2)also this issue more times occurring after update to latest version.
some time same video play with url (open with Auto play false)
and some time same video don't play with url and returning this error (open with Auto play)

[AVProVideo] Error: Loading failed. File not found, codec not supported, video resolution too high or insufficient system resources.

**Comment by rikensonaniNB** [comment]
i have deep check with memory profile and i found the memory related issue 
1)when i play video then allocate memory (for example: start memory is 0 and after play 0.35mb then i close video memory is 0.1mb to 0.2mb not releasing proper.) some time release proper and some time not release proper difference between starting memory and after close video

**Comment by Ste-RH** [comment]
> > 2)also this issue more times occurring after update to latest version. some time same video play with url (open with Auto play false) and some time same video don't play with url and returning this error (open with Auto play)
> > [AVProVideo] Error: Loading failed. File not found, codec not supported, video resolution too high or insufficient system resources.
> 
> thank you for support what about this error ? 2)also this issue more times occurring after update to latest version. some time same video play with url (open with Auto play false) and some time same video don't play with url and returning this error (open with Auto play)
> 
> [AVProVideo] Error: Loading failed. File not found, codec not supported, video resolution too high or insufficient system resources.

Does this happen when you have no other videos open?  If so, then it is likely just a failure to open due to lack of resources.

Can you reproduce this in our Demo_MediaPlayer scene?  Can you send over a full logcat when this happens?

**Comment by Ste-RH** [comment]
> i have deep check with memory profile and i found the memory related issue 1)when i play video then allocate memory (for example: start memory is 0 and after play 0.35mb then i close video memory is 0.1mb to 0.2mb not releasing proper.) some time release proper and some time not release proper difference between starting memory and after close video

This will be down to garbage collection.  You either need to force/action GC, or wait until the system cleans up on it's own.

We regularly run stress tests on AVPro Video releases.  I ran one on v2.6.5 on Android recently.  I stopped the test after I had 'open/played/closed...repeat' 100k times.  This is one of our unit tests for memory leaking issues in the asset.

**Comment by Ste-RH** [comment]
Does it happen ever when you first open the video after running the app?

**Comment by Chris-RH** [comment]
Does it happen when you first open a video after running the app?
Can you provide a full logcat please?

**Comment by Chris-RH** [comment]
Closed due to lack of reply
